### DISCUSSION TITLE
Modular dashboard

### DISCUSSION THREAD

**Discussion by skyrowl** [lunarphp/lunar#1051] [Feature Requests]
**Describe the solution you'd like**
Make the dashboard modular so that you can register widgets you'd like to display similar to how media transformations are 

**Additional context**
Per widget permissions would be a good point to manage what staff can see
Drag and drop available widgets to reorder them and being able to hide/display what users want to see on their first page would make it perfect

**Comment by skyrowl** [comment]
I was thinking of putting default values that could be overriden in config, something like
```json
// Is the widget enabled (0, 1, true, false)
"enabled": 1,
// Default order sorted by priority 0-99+
"order": 3,
// Min and max width (colspan xl/lg: 1-12, md: 1-8, sm: 1-6) accepting ranges defaults falling to 1 for min and max colspan for max. 
"width": {
  // sizable
  "xl": {
    "min": 4,
    "max": 8
  },
  // sizable
  "lg": {
    "min": 4,
    "max": 6
  },
  // fixed
  "md": 4,
  // fixed
  "sm": 3
}
````
When resizing we consider that the order is kept and we let grids do their magic while defined size will be stored using cookies to keep these values. 

Sizing them within the UI should not be a big deal as creating a little script that is included within the widget blade with a `@once` tag or include this in a widget layout used for every elements nested in the default hub layout on the dashboard. 

Updating widgets with the date (maybe other parameters in the future?) changing could be done in two ways:
- the easy but inefficient one would be to put a wire key with the current time on each components so that when updating the date, livewire refresh the page and so the keys that will regenerate components. Tho it will not be possible to make static widgets without updating them (as in filament "Hello, $user")
- the ""hard"" way is to setup a global livewire emit on the main component so that every child components _can_ update datas. Considering this most good charts libraries are able to refresh the charts with a fresh dataset through functions like `.refresh()`. In that way you can just entangle datas and refresh on update

Hope I answered, tell me if I forgot anything or if you need more explanations.

**Comment by glennjacobs** [comment]
I think this is likely to be more interesting when Livewire 3 comes out and we have lazy loading components. Some of the widgets could be reporting on intensive queries and it someone puts a whole bunch of them on their dashboard it could take a fair time to load.
